File: legacy/src/main/rs/emulate/legacy/config/varbit/DefaultBitVariableDefinition.java
package rs.emulate.legacy.config.varbit;

import rs.emulate.legacy.config.DefaultConfigDefinition;
import rs.emulate.legacy.config.SerializableProperty;
import rs.emulate.shared.util.DataBuffer;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;

import static rs.emulate.legacy.config.varbit.BitVariableProperty.VARIABLE;

/**
 * A default {@link BitVariableDefinition} used as a base for an actual definition.
 *
 * @param <T> The type of {@link BitVariableDefinition} this default is for.
 * @author Major
 */
public class DefaultBitVariableDefinition<T extends BitVariableDefinition> extends DefaultConfigDefinition<T> {

	@Override
	protected Map<Integer, SerializableProperty<?>> init() {
		Map<Integer, SerializableProperty<?>> properties = new HashMap<>(1);

		BiConsumer<DataBuffer, Variable> encoder = (buffer, bits) -> buffer.putShort(bits.getVariable())
				.putByte(bits.getHigh()).putByte(bits.getLow());

		Function<DataBuffer, Variable> decoder = buffer -> {
			int variable = buffer.getUnsignedShort();
			int low = buffer.getUnsignedByte();
			int high = buffer.getUnsignedByte();

			return new Variable(variable, high, low);
		};

		properties.put(1, new SerializableProperty<>(VARIABLE, Variable.EMPTY, encoder, decoder,
				Short.BYTES + 2 * Byte.BYTES, input -> Optional.empty())); // FIXME parser

		return properties;
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/config/varbit/DefaultBitVariableDefinition.kt
package rs.emulate.legacy.config.varbit

import rs.emulate.legacy.config.DefaultConfigDefinition
import rs.emulate.legacy.config.SerializableProperty
import rs.emulate.legacy.config.varbit.BitVariableProperty.Varbit
import rs.emulate.shared.util.DataBuffer

/**
 * A default [BitVariableDefinition] used as a base for an actual definition.
 *
 * @param <T> The type of [BitVariableDefinition] this default is for.
 */
class DefaultBitVariableDefinition<T : BitVariableDefinition> : DefaultConfigDefinition<T>() {

    override fun init(): Map<Int, SerializableProperty<*>> {
        val encoder = { buffer: DataBuffer, bits: Variable ->
            buffer.putShort(bits.variable)
                .putByte(bits.high)
                .putByte(bits.low)
        }

        val decoder = { buffer: DataBuffer ->
            val variable = buffer.getUnsignedShort()
            val low = buffer.getUnsignedByte()
            val high = buffer.getUnsignedByte()

            Variable.create(variable, low, high)
        }

        return hashMapOf(
            1 to SerializableProperty(Varbit, Variable.EMPTY, encoder, decoder,
                java.lang.Short.BYTES + 2 * java.lang.Byte.BYTES
            )
        )
    }

}
